src/main/java/com/github/minecraftschurlimods/arsmagicalegacy/common/entity/AbstractBoss.java

```diff
@@ -0,0 +1,296 @@
+                        for (int z = -width; z < width; z++) {
+                            level.destroyBlock(new BlockPos(getX() + x, getY() + y, getZ() + z), true, this);
+                        }
+                    }
+                }
+            }
+            return false;
+        }
+        SoundEvent sound = getHurtSound(pSource);
+        if (sound != null) {
+            level.playSound(null, this, sound, SoundSource.HOSTILE, 1f, 0.5f + random.nextFloat() * 0.5f);
+        }
+        return super.hurt(pSource, pAmount);
+    }
+
+    @Override
+    public void kill() {
+        hurtTime = 10;
+        remove(Entity.RemovalReason.KILLED);
+        bossEvent.setProgress(0);
+        bossEvent.setVisible(false);
+        bossEvent.removeAllPlayers();
+        updatePlayers();
+    }
+
+    @Override
+    public boolean canCastSpell() {
+        return action == Action.IDLE;
+    }
+
+    @Override
+    public boolean isCastingSpell() {
+        return action == Action.CAST;
+    }
+
+    @Override
+    public void setIsCastingSpell(boolean isCastingSpell) {
+        if (isCastingSpell) {
+            setAction(Action.CAST);
+        } else if (action == Action.CAST) {
+            setAction(Action.IDLE);
+        }
+    }
+
+    @Override
+    public AnimationFactory getFactory() {
+        return factory;
+    }
+
+    /**
+     * @return The attack sound of this boss, or null if no attack sound should be played.
+     */
+    @Nullable
+    public SoundEvent getAttackSound() {
+        return null;
+    }
+
+    /**
+     * @return The current action of this boss.
+     */
+    public Action getAction() {
+        return action;
+    }
+
+    /**
+     * Sets the current action.
+     *
+     * @param action The action to set.
+     */
+    public void setAction(Action action) {
+        this.action = action;
+        ticksInAction = 0;
+        level.broadcastEntityEvent(this, action.id);
+    }
+
+    /**
+     * @return The amount of ticks the boss is already using the current action.
+     */
+    public int getTicksInAction() {
+        return ticksInAction;
+    }
+
+    private void updatePlayers() {
+        if (!level.isClientSide()) {
+            Set<ServerPlayer> newSet = new HashSet<>();
+            for (ServerPlayer player : ((ServerLevel) level).getPlayers(EntitySelector.ENTITY_STILL_ALIVE.and(EntitySelector.withinDistance(0, 128, 0, 192)))) {
+                bossEvent.addPlayer(player);
+                startSeenByPlayer(player);
+                newSet.add(player);
+            }
+            Set<ServerPlayer> oldSet = Sets.newHashSet(bossEvent.getPlayers());
+            oldSet.removeAll(newSet);
+            for (ServerPlayer player : oldSet) {
+                bossEvent.removePlayer(player);
+                stopSeenByPlayer(player);
+            }
+        }
+    }
+
+    public enum Action {
+        IDLE(-1), CAST(-2), LONG_CAST(-3), SMASH(-4), SPIN(-5), STRIKE(-6), THROW(-7);
```

Comment:
Action enum in the abstract base class? please elaborate your decision

Reply:
1) Actions are shared between basically all bosses (IDLE and CAST are present on every boss, and all others are present on at least 3 bosses each iirc)
2) The id params are needed for syncing the animation states, which is handled in the abstract base class as well (because it would be the same everywhere)

Reply:
it is cleaner code if every boss had its own enum since they are not shared across **ALL** bosses (the id is fine)
Suggestion: make the ids constants and reuse them in the enum but have an enum per boss **or** just use the ids without an enum (less clean)

Reply:
Looking further into this, it is even more difficult to do than expected because the goals use the actions as well.